website/blog/2024-12-04-from-natural-language-to-sql/index.md

```diff
@@ -0,0 +1,1067 @@
+#### Step 4: Create and Store Embeddings in FAISS Vector Store
+Finally, we create and store the embeddings using FAISS. The FAISS.from_texts method takes all the chunks, computes their embeddings, and stores them in a high speed searchable vector database. This searchable database allows the engine to efficiently retrieve relevant SQL guidelines, significantly improving the success rate of executable SQL query generation.
+
+```python
+import os
+from langchain_community.vectorstores import FAISS
+from langchain_openai import OpenAIEmbeddings
+from langchain_community.document_loaders.recursive_url_loader import RecursiveUrlLoader
+from bs4 import BeautifulSoup as Soup
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+
+def setup_vector_store():
+    """Setup or load the vector store."""
+    if not os.path.exists('data'):
+        os.makedirs('data')
+
+    vector_store_dir = 'data/vector_store'
+
+    if os.path.exists(vector_store_dir):
+        # Load the vector store from disk
+        print("Loading vector store from disk...")
+        vector_store = FAISS.load_local(
+            vector_store_dir,
+            OpenAIEmbeddings(),
+            allow_dangerous_deserialization=True
+        )
+    else:
+        print("Creating new vector store...")
+        # Load SQL documentation
+        url = "https://www.w3schools.com/sql/"
+        loader = RecursiveUrlLoader(
+            url=url, max_depth=2, extractor=lambda x: Soup(x, "html.parser").text
+        )
+        docs = loader.load()
+
+        # Split documents into chunks
+        text_splitter = RecursiveCharacterTextSplitter(
+            chunk_size=500,
+            chunk_overlap=50,
+            separators=["\n\n", "\n", ".", "!", "?", ",", " ", ""],
+        )
+
+        documents = []
+        for doc in docs:
+            splits = text_splitter.split_text(doc.page_content)
+            for i, split in enumerate(splits):
+                documents.append({
+                    "content": split,
+                    "metadata": {"source": doc.metadata["source"], "chunk": i}
+                })
+
+        # Compute embeddings and create vector store
+        embedding_model = OpenAIEmbeddings()
+        vector_store = FAISS.from_texts(
+            [doc["content"] for doc in documents],
+            embedding_model,
+            metadatas=[doc["metadata"] for doc in documents]
+        )
+
+        # Save the vector store to disk
+        vector_store.save_local(vector_store_dir)
+        print("Vector store created and saved to disk.")
+
+    return vector_store
+```
+
+### SQLite Database
+
+The SQLite database is a key component of the Multilingual Query Engine, serving as the structured data repository that supports SQL query efficient generation, validation and execution by enabling:
```

Comment:
It might be worth mentioning why this is chosen for this example. Personally, I'm a huge fan of the portability and performance of sqlite and it definitely has many uses far beyond just demonstrations of concepts. As a self-contained data storage layer for an application, it's phenomenal at what it does and can greatly simplify developers' lives who would otherwise assume that they need to spin up a MySQL / PostGres DB for something that a local disk DB would handle much better.

Reply:
I’ve added a section detailing why SQLite was chosen for this example. I highlighted its lightweight, portable, and self-contained design, which, along with its ease of use and strong performance, makes it an ideal choice for this application.